src/Rector/ClassMethod/MakeEntitySetterNullabilityInSyncWithPropertyRector.php

```diff
@@ -102,18 +109,62 @@ public function refactor(Node $node): ?Node
         }
 
         $phpDocInfo = $this->phpDocInfoFactory->createFromNodeOrEmpty($property);
-
         if (! $phpDocInfo->hasByAnnotationClass('Doctrine\ORM\Mapping\ManyToOne')) {
             return null;
         }
 
         $param = $node->params[0];
 
-        /** @var NullableType $paramType */
+        if (! $this->hasNullableJoinColumn($phpDocInfo)) {
+            // remove nullable if has one
+            $paramType = $param->type;
+            if (! $paramType instanceof NullableType) {
+                return null;
+            }
+
+            $param->type = $paramType->type;
+            return $node;
+        }
+
+        // add nullable if does not have one
         $paramType = $param->type;
 
-        $param->type = $paramType->type;
+        // already nullable, lets skip it
+        if ($paramType instanceof NullableType) {
+            return null;
+        }
+
+        // we skip complex type as multiple or nullable already
+        if ($paramType instanceof ComplexType) {
+            return null;
+        }
+
+        // no type at all, there is nothing we can do
+        if (! $paramType instanceof Node) {
+            return null;
+        }
+
+        $param->type = new NullableType($paramType);
 
         return $node;
     }
+
+    private function hasNullableJoinColumn(PhpDocInfo $phpDocInfo): bool
+    {
+        // by default, the property is nullable if JoinColumn is missing
+        $joinColumnDoctrineAnnotationTagValueNode = $phpDocInfo->getByAnnotationClass(
+            'Doctrine\ORM\Mapping\JoinColumn'
+        );
+        if (! $joinColumnDoctrineAnnotationTagValueNode instanceof DoctrineAnnotationTagValueNode) {
+            return true;
+        }
+
+        $nullableNode = $joinColumnDoctrineAnnotationTagValueNode->getValue('nullable');
+        if (! $nullableNode instanceof ConstExprNode) {
+            // if the nullable explicit value is missing, by default the property is nullable
+            return true;
+        }
+
+        return $nullableNode instanceof ConstExprTrueNode;
+    }
```

Comment:
The deciding logic if the property should be made nullable,
or unmade nullable, was quite unclear. So I refactored it to this `hasNullableJoinColumn()` method.

Reply:
@alexander-schranz Could you review this logic makes sense now? :)
I do not use Doctrine, so I've only guessed it by test fixtures.